cmd/libsnap-confine-private/mountinfo.c

```diff
@@ -126,42 +127,110 @@ static void show_buffers(const char *line, int offset,
 	fprintf(stderr, ">%s<\n", line);
 
 	fputc('>', stderr);
-	for (int i = 0; i < strlen(line); ++i) {
+	for (size_t i = 0; i < strlen(line); ++i) {
 		int c = entry->line_buf[i];
 		fputc(c == 0 ? '@' : c == 1 ? '#' : c, stderr);
 	}
 	fputc('<', stderr);
 	fputc('\n', stderr);
 
 	fputc('>', stderr);
-	for (int i = 0; i < strlen(line); ++i)
+	for (size_t i = 0; i < strlen(line); ++i)
 		fputc('=', stderr);
 	fputc('<', stderr);
 	fputc('\n', stderr);
 #endif				// MOUNTINFO_DEBUG
 }
 
+static bool is_octal_digit(char c)
+{
+	return c >= '0' && c <= '7';
+}
+
 static char *parse_next_string_field(sc_mountinfo_entry * entry,
-				     const char *line, int *offset)
+				     const char *line, size_t * offset)
 {
-	int offset_delta = 0;
-	char *field = &entry->line_buf[0] + *offset;
-	if (line[*offset] == ' ') {
-		// Special case for empty fields which cannot be parsed with %s.
-		*field = '\0';
-		*offset += 1;
-	} else {
-		int nscanned =
-		    sscanf(line + *offset, "%s%n", field, &offset_delta);
-		if (nscanned != 1)
-			return NULL;
-		*offset += offset_delta;
-		if (line[*offset] == ' ') {
-			*offset += 1;
+	const char *input = &line[*offset];
+	char *output = &entry->line_buf[*offset];
+	const char *s;		// Used for escape logic.
+	size_t input_idx = 0;
+	size_t output_idx = 0;
+
+	// Scan characters until we run out of memory to scan or we find a
+	// space.  The kernel uses simple octal escape sequences for the
+	// following: space, tab, newline, backwards slash. Everything else is
+	// copied verbatim.
+	for (;;) {
+		int c = input[input_idx];
+		if (c == '\0') {
+			// The string is over before we see anything then
+			// return NULL. This is an indication of end-of-input
+			// to the caller.
+			if (output_idx == 0) {
+			    return NULL;
+			}
+			// The scanned line is NUL terminated. This ensures that the
+			// terminator is copied to the output buffer.
+			output[output_idx++] = '\0';
+			// NOTE: we are no longer advancing the reading index
+			// because we reached the end of the buffer.
+			break;
+		} else if (c == ' ') {
+			// Fields are space delimited or end-of-string terminated.
+			// Represent either as the end-of-string marker, skip over it,
+			// and stop parsing.
+			output[output_idx++] = '\0';
```

Comment:
Same here, we don't use output_idx after we break out of the loop, so no need to increment.

Reply:
Indeed, I only did this because all the code that assigns to output already increments output_idx, still you are correct here and above. Both increments can be dropped.